30 update codebase to work with Python 3.8 (and thus ase v3.23)

logsdail opened the pull request:
Testing to updated codebase with ASE release

logsdail commented:
Noting here that the docs runs with Python 3.9 - should we standardise to this version, perhaps?

logsdail commented:
So I think after discussion today at a Group meeting, this has become an "update Python version to the most suitable for our HPC architectures". 

Current Python compatibility of code: Python 3.7 (i.e., CI tests work to this version)

Current HPC system Python provisions:
Hawk: 3.7 (used currently, default), 3.9.2, 3.10.4, 3.11.2
Isambard: 3.7.3.2 (used currently), 3.8.5.1 (default)
Young: 3.6 (used currently), 3.7.0, 3.7.2, 3.7.4, 3.8.0, 3.8.6 (default), 3.9.0, 3.9.1, 3.9.6, 3.9.10, 3.11.3, 3.11.4
ARCHER2: 3.6.15 (used currently) 3.9.13.1 (default) 3.10.10

logsdail commented:
> So I think after discussion today at a Group meeting, this has become an "update Python version to the most suitable for our HPC architectures".
> 
> Current Python compatibility of code: Python 3.7 (i.e., CI tests work to this version)
> 
> Current HPC system Python provisions: Hawk: 3.7 (used currently, default), 3.9.2, 3.10.4, 3.11.2 Isambard: 3.7.3.2 (used currently), 3.8.5.1 (default) Young: 3.6 (used currently), 3.7.0, 3.7.2, 3.7.4, 3.8.0, 3.8.6 (default), 3.9.0, 3.9.1, 3.9.6, 3.9.10, 3.11.3, 3.11.4 ARCHER2: 3.6.15 (used currently) 3.9.13.1 (default) 3.10.10

My take home from this is the minimum we should go to is 3.8 or 3.9, but certainly not towards the high numbers.

logsdail commented:
@PavelStishenko as suggested I've put a tag on the mainline version, and once merged we'll bumpy the minor version number.

ikowalec commented on examples/analyse_forces.py:
```diff
@@ -28,7 +34,7 @@ def test_analyse_forces():
     print(is_converged(atoms, fmax))
     assert not is_converged(atoms, fmax)
-    print(atoms.calc.forces)
+    print(atoms.get_forces())
```
`atoms.get_forces()` will calculate the forces, previously, the `atoms.calc.forces` retrieved an empty list from the results dictionary of the calculator object

ikowalec replied:
I suggest removing the print statement completely

logsdail replied:
Fair - and solution agreed.

ikowalec commented on .github/workflows/main.yml:
```diff
@@ -15,38 +15,45 @@ jobs:
+        python -m pip install --upgrade pymatgen
         
+        # Debug
+        python -m pip freeze
```
Do we want to keep the debug?

logsdail replied:
I think no harm to keep it - means we can see versions if any questions arise, and low cost.

ikowalec commented on examples/analyse_counterpoise.py:
```diff
@@ -1,31 +1,43 @@
+#!/usr/bin/env python3
 '''
 
 TODO: Needs high level description
```
The description is required, but technically an issue outside of the goal of this pull request.

ikowalec commented on examples/build_symmetry_mirror.py:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python3
 '''
-This example shows usage of symmetry operation tools for FCC lattice.
+This example shows usage of symmetry operation utils for FCC lattice.
```
not sure, about this change since utility functions reside in the `utils` folder

ikowalec commented on carmm/utils/python_env_check.py:
```diff
@@ -0,0 +1,71 @@
+
+@author Firas Assa
+
+deprecation() is a sub function of python_env_check() that outputs a DeprecationWarning
```
Can the descriptions be moved to the respective functions?

ikowalec commented on carmm/run/workflows/react.py:
```diff
@@ -287,6 +288,8 @@ def get_mulliken_charges(self, atoms: Atoms, restart: bool = True):
 
         counter, out, subdirectory_name, initial = helper.restart_setup()
 
+        print(out)
```
Presumably leftover from the debugging process

logsdail replied:
Removed.

ikowalec reviewed (approved):
There is some tidy up to do, but that is not related to the compatibility issues, which have been resolved successfully.

logsdail commented on examples/analyse_counterpoise.py:
```diff
@@ -1,31 +1,43 @@
+        toy_calc.set(spin='collinear', relativistic='atomic_zora scalar')
+    else:
+        from ase.calculators.aims import AimsProfile, Aims
+        print('3.23.0')
```
No - removing

logsdail commented on carmm/utils/python_env_check.py:
```diff
@@ -0,0 +1,71 @@
+    # sub-function of python_env_check()
+    warnings.warn(message, DeprecationWarning, stacklevel=2)
+
+def ase_env_check(version=None):
```
Resolved with new commit

logsdail commented:
@ikowalec thanks for comments - all addressed. Does someone else want to do a second review, please?

logsdail commented:
@GaryLZW I simplified the `analyse_counterpoise` test, as the checking for ASE now happens in `get_aims_calculator`. The CI tests pass but you might want to check if the code works correctly for further use cases.

GaryLZW commented on carmm/run/aims_calculator.py:
```diff
@@ -21,29 +21,44 @@ def get_aims_calculator(dimensions, k_grid=None, xc="pbe", compute_forces="true"
+        parameter_dict['relativistic'] = ('atomic_zora', 'scalar')
+
+    fhi_calc = Aims(
+        profile=AimsProfile(command='xc'),
```
Hi Andy. I am not sure about this line. My understanding is this.
If ase version is 3.22, `AimsProfile(command='xc')` is an Aims calculator. Will this pollute the `control.in` file `fhi_calc` generates? I didn't find where they remove the non-existing parameters in the old write_control function.
If ase version is 3.23, `AimsProfile(command='xc')` provides a profile for `fhi_calc`. This seems to let dry run work. But does it prevent the `fhi_calc` to load commands and paths when doing real calculations?

logsdail replied:
Fair questions, but hard to answer as you are basically asking for situations outside the boundaries of our CI tests. The only way we can find out is by testing, and then improving the CI tests if things don't work.

logsdail replied:
For actual use with ASE v3.23, all we have currently is the CI tests - we'd have to deploy and use to see if it works, or doesn't, and then resolve as necessry.

logsdail replied:
(We can't fix bugs we don't know about!)

logsdail replied:
@GaryLZW is this at least clearer now in implementation? And we can work to see how things work when actually deployed?

GaryLZW replied:
Hi Andy. Yes. The code looks clearer to me now. Happy to resolve further issues after deploying.

GabrielBram replied:
I think a more permanent solution should be feasible when we get around to re-working ReactAims and related calculator handling routines for ASE 3.23 - I can imagine we can create a dummy profile that is loaded for CI testing, and loads the user-generated profile otherwise.

logsdail replied:
Agreed - there is a better solution out there. I think we also need to make the executable commands internal, rather than system environment settings.

GaryLZW commented:
> @GaryLZW I simplified the `analyse_counterpoise` test, as the checking for ASE now happens in `get_aims_calculator`. The CI tests pass but you might want to check if the code works correctly for further use cases.

This looks fine to me^^

FirasAssa commented:
During my project the following changes were made:
The err_handler folder was removed and a utils folder was created in its place
A python_env_check file was created with python_env_check() and is_env_python() inside
Implemented python_env_check() in anylyse_forces example 
Implemented python_env_check() in analyse_calculator
Restricted mace-torch to 0.3.4, because it was failing on v3.7 
Updated CatLearn to the most recent version to fix import error in mlneb.py, by getting the version of CatLearn directly from Git 

Thinks still to do after the project:
In python v3.8 ci-test there are 4 places where .BadConfiguration Error occurs. Presumably it is the same type of problem that can be fixed in one go
Errors:
ERROR analyse_counterpoise.py - ase.calculators.calculator.BadConfiguration: No configuration of aims
ERROR run_aims.py - ase.calculators.calculator.BadConfiguration: No configuration of aims
ERROR run_workflows_ReactAims.py - ase.calculators.calculator.BadConfiguration: No configuration of aims
ERROR run_workflows_ReactAims_parallel.py - ase.calculators.calculator.BadConfiguration: No configuration of aims

logsdail commented:
[like]  Andrew Logsdail reacted to your message:
________________________________
From: FirasAssa ***@***.***>
Sent: Sunday, August 4, 2024 9:13:06 PM
To: logsdail/carmm ***@***.***>
Cc: Andrew Logsdail ***@***.***>; State change ***@***.***>
Subject: Re: [logsdail/carmm] 30 update codebase to work with Python 3.8 (and thus ase v3.23) (PR #163)

External email to Cardiff University - Take care when replying/opening attachments or links.
Nid ebost mewnol o Brifysgol Caerdydd yw hwn - Cymerwch ofal wrth ateb/agor atodiadau neu ddolenni.



During my project the following changes were made:
The err_handler folder was removed and a utils folder was created in its place
A python_env_check file was created with python_env_check() and is_env_python() inside
Implemented python_env_check() in anylyse_forces example
Implemented python_env_check() in analyse_calculator
Restricted mace-torch to 0.3.4, because it was failing on v3.7
Updated CatLearn to the most recent version to fix import error in mlneb.py, by getting the version of CatLearn directly from Git

Thinks still to do after the project:
In python v3.8 ci-test there are 4 places where .BadConfiguration Error occurs. Presumably it is the same type of problem that can be fixed in one go
Errors:
ERROR analyse_counterpoise.py - ase.calculators.calculator.BadConfiguration: No configuration of aims
ERROR run_aims.py - ase.calculators.calculator.BadConfiguration: No configuration of aims
ERROR run_workflows_ReactAims.py - ase.calculators.calculator.BadConfiguration: No configuration of aims
ERROR run_workflows_ReactAims_parallel.py - ase.calculators.calculator.BadConfiguration: No configuration of aims

—
Reply to this email directly, view it on GitHub<https://github.com/logsdail/carmm/pull/163#issuecomment-2267678961>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFMK4MYUDKHIE4IDSR2UDILZP2KOFAVCNFSM6AAAAABIWCQXZCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENRXGY3TQOJWGE>.
You are receiving this because you modified the open/close state.Message ID: ***@***.***>
